torch/sparse/__init__.py

```diff
@@ -150,6 +150,48 @@
 """)
 
 
+spmm_reduce = _add_docstr(_sparse.spmm_reduce, r"""
+sparse.spmm_reduce(self, other, reduce, *, row_indices=None, ccol_indices=None, csr2csc=None) -> Tensor
+
+Reduce rows from dense matrices :attr:`other` at the locations specified by the sparsity pattern of :attr:`self`.
+The matrix :attr:`self` is not added to the final result.
+
+For each element in :attr:`self` at index [i, j], pick the rows from :attr:`other` specified by j,
+and reduce to :attr:`output` at row i.
+
+.. note::
+    :attr:`self` must be a sparse CSR tensor. :attr:`other` must be a dense tensors.
+    This function is implemented only for tensors on CPU device.
+
+Args:
+    self (Tensor): a sparse CSR matrix of shape `(m, n)`.
+    other (Tensor): a dense matrix of shape `(n, k)` to be reduced.
+    reduce (str): the reduction operation to apply for non-unique indices
+        (:obj:`"sum"`, :obj:`"mean"`, :obj:`"amax"`, :obj:`"amin"`).
+
+Keyword args:
+    row_indices (Tensor, optional): row indices for each element in :attr:`input`, 1-D tensor of size `nse`.
+    ccol_indices (Tensor, optional): compressed col indices for each element in :attr:`input`, 1-D tensor of size `n + 1`.
+    csr2csc (Tensor, optional): permute pattern to convert :attr:`input` from CSR to CSC.
```

Comment:
We have support for the CSC layout. If you do `csr_tensor.transpose()` you'll get a CSC Tensor. We could just register another kernel for this function under SparseCSC instead of adding this argument?

Reply:
the proposed operator `spmm_reduce` is a direct mapping of a PyG operator at https://github.com/rusty1s/pytorch_sparse/blob/master/csrc/spmm.cpp#L55. This used for information aggregation and PyG has custom module called `SparseTensor` (which is not the same thing as `torch.Tensor` with a SparseCsr layout), it cached some meta info for backward, e.g. `row_indices, ccol_indices and csr2csc`.

Right now we do not expect that this operator will be directly used by user, it is more likely that user are directly using PyG (yet the underlying optimization comes from torch). So for the current stage, it is more convenient to keep the API aligned with PyG.

We are also upgrading on the PyG side to align `SparseTensor` to `torch.Tensor (with SparseCsr layout)`, so perhaps we can put this for the next step.